### DISCUSSION TITLE
Servo deck - an easy way to add a servo to your Crazyflie

### DISCUSSION THREAD

**Discussion by matejkarasek** [bitcraze/discussions#1208] [Show and tell]
# Servo deck
Many people (including us) would like to add a servo to their Crazyflies (or Flappers...) e.g. for a gripper. Based on several discussions here (#1005, #729, ...), we recently implemented a "servo_deck" (https://github.com/bitcraze/crazyflie-firmware/pull/1340), which allows to connect a **single** servo to your Crazyflie and control the servo's position via the parameters' framework. 

It is not a hardware deck, but rather a driver, so you may need to do some soldering in combination with the prototyping or breakout decks, or use the BigQuad deck. Still, this should make integration of servos much easier.

In any case, this deck is experimental and you should be aware that you may damage your Crazyflie, so please proceed with caution and only if you know what you are doing.

## How do I enable the deck?

To enable this functionality, you have to force "bcServo" by adding the following lines in your defconfig:
```
CONFIG_DECK_FORCE="bcServo"`
CONFIG_DECK_SERVO=y
```
If you need to force multiple decks, just separate them by a colon, e.g. `CONFIG_DECK_FORCE="bcServo:bcUSD"`

## Where do I connect the servo?

By default, the deck assumes the servo signal is connected to TX2/PA2. You can also change the mapping to a different pin to avoid conflicts with other decks you may also like to use. 
The available options are TX2 (default), IO2, IO3, RX2, MOSI and IO1 (this last option only works on Bolt 1.1). 

To use a non default pin, add an extra line in the defconfig, e.g.:
```
CONFIG_DECK_SERVO_USE_RX2=y
```

To power the servo, on the standard Crazyflie 2.1 you can use the VCOM and GND, this will power the servo directly from your battery (max 4.2 V). This also works on the Bolt, where VCOM is 5V when powered by 2S or higher-voltage battery, just pay attention that VCOM current is then limited to 400mA (which could be reached easily when the servo is stalled…).

You can use the expansion or prototyping decks and solder the servo wires to it.
Another option is to use the BigQuad deck where servo-compatible connectors are already installed (M1 -> TX2, M2 -> IO3, M3 -> IO2, M4 -> RX2, or even CPPM -> MOSI) and you just need to provide power to the deck. In this case, you may need to disable the OW memory of the Big Quad deck such that the Big Quad deck driver is not initiated or adjust the firmware to only force the Servo deck driver.

## Servo in power distribution

This PR is meant for "addons" and so it is not meant to be used in the power distribution. If you need a servo that will be used for attitude control, please check the flapper platform for inspiration: https://github.com/bitcraze/crazyflie-firmware/pull/1132

**Comment by knmcguire** [comment]
Awesome! thanks for sharing this in the community 😄

**Comment by WillWu88** [comment]
Thanks for all the hard work! So excited that this is now officially a feature.

**Comment by aasignacion** [comment]
Hello @matejkarasek 

Thank you for sharing this.
I am trying to follow your instructions to add servo.

Where and How can I set the desired angle?

I was thinking that I can integrate in the controller_pid.c by including servo.h. Then, call the servoInit() in controllerPidInit(), call servoTest() to controllerPidTest() and use the function servoSetAngle() in the controllerPid()?
I am just guessing.

Could you guide me so that I can add it the Flapper drone?
Thank you.

**Reply by matejkarasek** [reply]
Hello @aasignacion 

So first you need to enable the deck by adding `CONFIG_DECK_FORCE="bcServo"`,
`CONFIG_DECK_SERVO=y`, and the pin configuration e.g. `CONFIG_DECK_SERVO_USE_IO1` (the default pin is already used on the flapper) in your flapper_defconfig file. This will make sure the driver gets initialized, no need to call servoInit().

Then you just include servo.h in the file, where you would like to call servoSetAngle(). 
You can adjust the limits of the servo via persistent parameters of the "servo" parameter group.

**Reply by aasignacion** [reply]
Thank you so much.
One more question.
How can I use it in python code?

**Reply by matejkarasek** [reply]
Next to the servoSetAngle function, you can also control the position of the servo be setting the "servo.servoAngle" parameter. 
This you can do easily from Python, check e.g. this example: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/#step-3-parameters

**Reply by aasignacion** [reply]
Thank you.
